### DISCUSSION TITLE
Analysis Engine - @types

### DISCUSSION THREAD

**Discussion by ansrlm** [githru/githru-vscode-ext#71] [General]
### 목적 및 상황
- @types/analysis-engine(가칭) 파일을 제공하는 방식에 대해 협의합니다.
- analysis-engine은 src/types하위에 type파일을 가지고 있습니다.
- npm packages monorepo형식을 채택하고 있어서 `@githru/${package}`로 접근이 가능합니다.
- 번들러 세팅이 되어 있어서 type.d.ts파일을 따로 꺼내줄 수 있습니다.
- 인터페이스는 보수적으로 생각할 경우 많은 변경사항이 존재할 수 있습니다.

### 타입 제공 방식
- monorepo에서 type을 꺼내오는 방식
  - @githru-vscode-ext/view의 package.json에 devDependency로 @githru-vscode-ext/analysis-engine을 targeting을 합니다.
  - `import type { DifferenceStatistic } from "@githru-vscode-ext/analysis-engine/src/types/CommitRaw";`
  - 우려되는 이슈
    - interface 정의에 대한 PR 머지 -> 해당 type파일을 사용하는 view repo에서 PR 머지 -> interface 정의 변경 PR 머지 -> 언제나 에러남
- npm publish를 하는 경우
  - monorepo에서 type파일을 직접 참조하는 것에서 생기는 문제 회피 가능
  - @githru-vscode-ext/view의 package.json에 devDependency로 배포한 analysis-engine을 targeting을 합니다.
  - 우려되는 이슈
    - 배포 단위를 어떻게 가져갈 것인가?
      - major 0점대에서는 계속 배포해도 이슈는 없을 것 같음

### 타입 제공 형태
- 처리방식 1
  - src/types/index에서 하위의 파일들을 모두 가지고 있다가 export합니다.
  - type을 가져오는 가이드를 언제나 `@githru-vscode-ext/analysis-engine/src/types`만 하도록 가이드합니다.
  - 장점 : index.ts에서 타입 파일들에 대한 제어 가능
  - 단점 : index.ts에 대한 관리 필요성
  - https://github.com/ant-design/pro-components/blob/master/packages/components/src/index.tsx
- 처리방식 2
  - 각각 export합니다.
  - 장점 : 추가 작업이 필요 없음
  - 단점 : ?
  - https://github.com/facebookexperimental/Recoil/blob/main/typescript/index.d.ts

**Comment by hy57in** [comment]
처리방식1로 결정.
but 현재 시점에는 진행하지 않을 것 같습니다. 해당 파일이 아마 계속 conflict이 날 여지가 있어서요.
모든분들 다 기능에 대한 파일이 merge가 끝난 이후에 추가 PR을 올려서 처리할 예정입니다
